feat(api): Add 'count' parameter to return total item count in API v4

elisa-a-v opened the pull request:
Resolves #4506 

This PR introduces a new `count=on` query parameter to our paginated API endpoints. When this parameter is included in a request, the API returns only the total count of items matching the query filters, without including any result data.

If no `count=on` is added to the request, a count key is now included in the response just like in v3, but instead of providing the count immediately, it provides a URL that points to the count endpoint for the current query parameters to guide users on how to obtain the total count if needed.

V4 removed the count due to the overhead of counting huge querysets in each request, which is why we're adding this tweaked version of the count in the response so the users have access to the information if needed without reintroducing the overhead.

Also updated docs to reflect these changes:
![image](https://github.com/user-attachments/assets/569a224b-5128-425c-8096-984aa188a137)

mlissner reviewed (approved):
Looks nice! 

One thought that might be helpful is to confirm the number and type of queries via a test. Could you add that? 

The rest looks great to me, if it looks good to other folks.

mlissner commented:
Process notes:
 - I assigned this to Alberto since he worked in this area most recently. It's probably best if I continue doing this for PRs.
 - I set him as reviewer and removed Eduardo. It's probably best if I continue doing this for PRs.
 - I added it to the current sprint and put it in the To Do column. It's probably best if you do this in the future. :)

Thank you!

albertisfu commented on cl/api/templates/rest-docs-vlatest.html:
```diff
@@ -324,6 +324,30 @@ <h3 id="ordering">Ordering</h3>
+      <li>The response does not include any result data, which can improve performance for large datasets.</li>
+    </ul>
+    <p>In standard paginated responses, a <code>count</code> key is included with the URL to obtain the total count for your query:</p>
+    <pre class="pre-scrollable">curl "{% get_full_host %}{% url "opinion-list" version="v4" %}?cited_opinion=32239
```
Just a missing quote.
```suggestion
    <pre class="pre-scrollable">curl "{% get_full_host %}{% url "opinion-list" version="v4" %}?cited_opinion=32239"
```

elisa-a-v replied:
oof thanks for catching that!

albertisfu reviewed (commented):
This looks pretty good to me.

I just left a couple of minor comments.

I also agree with Mike's suggestion to add or tweak the current tests to verify the number and type of queries performed during a regular request and a count request. To ensure that no count query is executed during regular requests, and only the count query is executed during count requests. I think you can base on `ApiQueryCountTests`

Thank you!

mlissner commented on cl/api/pagination.py:
```diff
@@ -88,6 +90,12 @@ def paginate_queryset(self, queryset, request, view=None):
 
         self.version = request.version
         self.request = request
+        self.is_count_request = request.query_params.get("count") == "on"
```
It's OK, if it's in v3 too, but if it's this easy, sure, it's better not to bring hot new features to v3.

elisa-a-v commented:
> Process notes:
>
> - I assigned this to Alberto since he worked in this area most recently. It's probably best if I continue doing this for PRs.
> - I set him as reviewer and removed Eduardo. It's probably best if I continue doing this for PRs.
> - I added it to the current sprint and put it in the To Do column. It's probably best if you do this in the future. :)

That makes sense, so from now on I'll just add you as a reviewer once a PR is ready for review, and I'll add it to the current sprint 🫡

elisa-a-v commented:
Thank you @albertisfu and @mlissner for the comments! I'm working on the tests and I think the class you mentioned (`ApiQueryCountTests`) seems like the right place to add the new tests, do you agree?

albertisfu commented:
> I'm working on the tests and I think the class you mentioned (ApiQueryCountTests) seems like the right place to add the new tests, do you agree?

Yeah, I agree. that's a good place to add the new tests!

albertisfu reviewed (commented):
Thanks @elisa-a-v the changes and the API count tests added look great.

We can merge this one @mlissner
